core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java

```diff
@@ -125,22 +106,26 @@ public Argument build(Type expectedType, Object value, StatementContext ctx)
 
         ArgBuilder v = BUILDERS.get(expectedClass);
         if (v != null) {
-            return v.build(value);
+            return Optional.of(v.build(value));
         }
 
         // Enums must be bound as VARCHAR.
         if (value instanceof Enum) {
-            return STR_BUILDER.build((((Enum<?>)value).name()));
+            Enum<?> enumValue = (Enum<?>) value;
+            return Optional.of(STR_BUILDER.build(enumValue.name()));
         }
 
         if (value instanceof Optional) {
             Object nestedValue = ((Optional<?>)value).orElse(null);
             Type nestedType = findOptionalType(expectedType, nestedValue);
-            return ctx.argumentFor(nestedType, nestedValue);
+            return ctx.findArgumentFor(nestedType, nestedValue);
+        }
+
+        if (expectedType == Object.class) {
+            return Optional.of(OBJ_BUILDER.build(value));
         }
 
-        // Fallback to generic ObjectArgument
-        return OBJ_BUILDER.build(value);
```

Comment:
This was intentional.

Compare the old `Foreman.waffle` and the new `ArgumentRegistry.findArgumentFor`.

Foreman iterated the list once, and searched for a result as well as a fallback candidate in the same loop. Thus, BuildInArgumentFactory was always consulted last.

The new method is using streams (iterating once by exact argument type, then again if necessary by Object.class) and will return the first argument that comes out of the stream.

I added the if clause around OBJ_BUILDER, to prevent BuiltInArgumentMapper from returning a result on the first iteration, to allow each registered mapper a second chance to return a result for Object.class. A few tests actually depend on this.